mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala

```diff
@@ -269,6 +322,13 @@ class KMeans @Since("1.5.0") (
       .setMaxIterations($(maxIter))
       .setSeed($(seed))
       .setEpsilon($(tol))
+
+    if (isSet(initialModel)) {
+      require($(initialModel).parentModel.clusterCenters.length == $(k), "mismatched cluster count")
```

Comment:
This check should go in validateAndTransformSchema as well (or instead).  Also, print mismatched values in error message.